File: Data Structures & Algorithms/binary-search/submission-1.java
class Solution {
    private int recBinarySearch(int l, int r, int[] nums, int target) {
        if (l > r) return -1;

        int m = (l + r) / 2;

        if (nums[m] == target) return m;
        if (target > nums[m]) {
            return recBinarySearch(m + 1, r, nums, target);
        } else {
            return recBinarySearch(l, m - 1, nums, target);
        }
    }

    private int itrBinarySearch(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        int m;

        while (l <= r) {
            m = (l + r) / 2;
            if (nums[m] == target) return m;
            if (target > nums[m]) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }

        return -1;
    }

    public int search(int[] nums, int target) {
        // return recbinarySearch(0, nums.length - 1, nums, target);
        return itrBinarySearch(nums, target);
    }
}

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-1.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        cur_sum = 0
        prefix_dict = defaultdict(int)
        prefix_dict[0] = 1
        res = 0
        for n in nums:
            cur_sum += n
            target = cur_sum - k
            res += prefix_dict[target]
            prefix_dict[cur_sum] += 1
        
        return res
